### ISSUE TITLE
Benchmarking query times without sharding

### ISSUE THREAD

**Issue by jayavanth** [apple/swift-homomorphic-encryption#45]
I'm trying to benchmark query times for keyword PIR with and without sharding the database. Where should I specify the number of shards in this code? https://github.com/apple/swift-homomorphic-encryption/blob/main/Benchmarks/PrivateInformationRetrievalBenchmark/PrivateInformationRetrievalBenchmark.swift

I'm quite new to Swift so any hint might be helpful. I see that `KeywordDatabaseConfig` is being used in some places in other files. Do I have to use that here to control the sharding?

**Comment by fboemer** [comment]
The benchmarks don't support sharding. But you can just change the database to be the size of a single shard by adjusting the `entryCount` [here](https://github.com/apple/swift-homomorphic-encryption/blob/780e3215cb48a5c681286d8f41d8a982bc313ccf/Benchmarks/PrivateInformationRetrievalBenchmark/PrivateInformationRetrievalBenchmark.swift#L312-L317).

E.g., the benchmark today has a database of 10000 entries. If you want to simulate performance on 10 shards, just change the entryCount to 1000. (In practice, there will be some variance in shard size due to random sharding, but the average, e.g., 1000, should be a good approximation).

Or, if you prefer, you can run `PirProcessDatabase` (see https://swiftpackageindex.com/apple/swift-homomorphic-encryption/main/documentation/pirprocessdatabase) with `trialsPerShard > 0`, which will also log the server runtime for each shard.

**Comment by jayavanth** [comment]
I got an output like this for database of size 100k elements
```
2024-08-06T06:10:45+0000 info PIRProcessDatabase : [PIRProcessDatabase] Processing database with configuration: {
  "algorithm" : "mulPir",
  "cuckooTableConfig" : {
    "bucketCount" : {
      "allowExpansion" : {
        "expansionFactor" : 1.1,
        "targetLoadFactor" : 0.9
      }
    },
    "hashFunctionCount" : 2,
    "maxEvictionCount" : 100,
    "maxSerializedBucketSize" : 1280,
    "multipleTables" : true
  },
  "inputDatabase" : "coo.txtpb",
  "outputDatabase" : "output-coo-SHARD_ID.bin",
  "outputPirParameters" : "pir-coo-params-SHARD_ID.txtpb",
  "rlweParameters" : "n_4096_logq_27_28_28_logt_6",
  "sharding" : {
    "shardCount" : 1
  },
  "trialsPerShard" : 1
}
2024-08-06T06:10:45+0000 info PIRProcessDatabase : [PIRProcessDatabase] Sharded database into 1 shards
2024-08-06T06:10:45+0000 info PIRProcessDatabase : [PIRProcessDatabase] Processing shard 0
2024-08-06T06:10:46+0000 info PIRProcessDatabase : [PIRProcessDatabase] Validating shard 0
2024-08-06T06:10:49+0000 info PIRProcessDatabase : [PIRProcessDatabase] ValidationResults {
  evaluation key size : 28.2 KB (9 keys),
  noise budget : 5.5,
  query size : 28.2 KB (1 ciphertexts),
  response size : 26.7 KB (2 ciphertexts),
  runtime (ms) : [294.8]
}
```
does that mean it took 294.8 ms to get query one time from the server? Also wondering if there is an executable for running server and client like `PIRProcessDatabase`

**Comment by fboemer** [comment]
> does that mean it took 294.8 ms to get query one time from the server? 

Yes (though may wish to use `trialsPerShard > 1` to get a sense for the variance. See also the `Parameter Tuning` [docs](https://swiftpackageindex.com/apple/swift-homomorphic-encryption/documentation/privateinformationretrieval/parametertuning) for some hints to tune the performance.

> Also wondering if there is an executable for running server and client like PIRProcessDatabase

No; though https://github.com/apple/live-caller-id-lookup-example has a sample [PIRService](https://swiftpackageindex.com/apple/live-caller-id-lookup-example/main/documentation/pirservice) which may give some guidance.

**Comment by jayavanth** [comment]
Thank you! Is there a way to query the server in parallel? Sharding might not be an option for my application as I can't leak any information about the query

**Comment by fboemer** [comment]
> Thank you! Is there a way to query the server in parallel? Sharding might not be an option for my application as I can't leak any information about the query

Not today, though it should be relatively straightforward to parallelize the server compute in a few places:
* [here](https://github.com/apple/swift-homomorphic-encryption/blob/1cff02b950bfac8b572936169ceae8ef45e1af80/Sources/PrivateInformationRetrieval/MulPir.swift#L397-L398)
* [here](https://github.com/apple/swift-homomorphic-encryption/blob/1cff02b950bfac8b572936169ceae8ef45e1af80/Sources/PrivateInformationRetrieval/MulPir.swift#L351-L352)
* [here](https://github.com/apple/swift-homomorphic-encryption/blob/1cff02b950bfac8b572936169ceae8ef45e1af80/Sources/PrivateInformationRetrieval/PirUtil.swift#L128)

Feel free to open a new issue for this feature request.

**Comment by fboemer** [comment]
Closing, since the original question was addressed.
